prov/util/src/util_mr_monitor.c

```diff
@@ -0,0 +1,151 @@
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <ofi_mr.h>
+
+static struct ofi_mr_monitor global_mr_monitor;
+
+static inline int
+util_check_mr_monitor_attr_sanity(struct ofi_mr_monitor_attr *attr)
+{
+	/* callbacks and provider must be provided */
+	if (!attr || !attr->prov ||
+	    !attr->init || !attr->open || !attr->close ||
+	    !attr->subscribe || !attr->unsubscribe || !attr->get_event)
+		return -FI_EINVAL;
+
+	/* valid otherwise */
+	return FI_SUCCESS;
+}
+
+int ofi_mr_monitor_init(struct ofi_mr_monitor_attr *attr)
+{
+	int ret;
+
+	ret = util_check_mr_monitor_attr_sanity(attr);
+	if (ret)
+		return ret;
+
+	global_mr_monitor.attr = *attr;
+	fastlock_init(&global_mr_monitor.lock);
+	global_mr_monitor.ref_cnt = 0;
+	dlist_init(&global_mr_monitor.rq_head);
+
+	return global_mr_monitor.attr.init();
+}
+
+int ofi_mr_monitor_open(struct ofi_mr_monitor **mr_monitor)
+{
+	int ret = FI_SUCCESS;
+
+	fastlock_acquire(&global_mr_monitor.lock);
+	if (global_mr_monitor.ref_cnt == 0)
+		ret = global_mr_monitor.attr.open(&global_mr_monitor.notifier);
+
+	if (!ret)
+		global_mr_monitor.ref_cnt++;
+	fastlock_release(&global_mr_monitor.lock);
+
+	if (!ret)
+		*mr_monitor = &global_mr_monitor;
+
+	return ret;
+}
+
+int ofi_mr_monitor_close(struct ofi_mr_monitor *mr_monitor)
+{
+	int ret = FI_SUCCESS;
+
+	fastlock_acquire(&mr_monitor->lock);
+	if (mr_monitor->ref_cnt == 0)
+		ret = -FI_EINVAL;
+
+	if (mr_monitor->ref_cnt == 1)
+		ret = mr_monitor->attr.close(mr_monitor->notifier);
+
+	if (!ret)
+		mr_monitor->ref_cnt--;
+	fastlock_release(&mr_monitor->lock);
+
+	return ret;
+}
+
+static void util_mr_monitor_read_events(struct ofi_mr_monitor *mr_monitor)
+{
+	struct ofi_mr_monitor_context *context = NULL;
+	struct ofi_mr_monitor_rq *rq;
+	int ret, len = (int)sizeof(uint64_t);
+
+	do {
+		ret = mr_monitor->attr.get_event(mr_monitor->notifier,
+						 (void *)&context, len);
+		if (ret != len) {
+			FI_DBG(UTIL_MR_MONITOR_GET_PROV(mr_monitor), FI_LOG_MR,
+			       "no more events to be read\n");
+			break;
+		}
+
+		FI_DBG(UTIL_MR_MONITOR_GET_PROV(mr_monitor), FI_LOG_MR,
+		       "found event, context=%p rq=%p cookie=%lx\n",
+		       context, context->rq, context->cookie);
+
+		rq = context->rq;
+		fastlock_acquire(&rq->lock);
+		dlist_insert_tail(&context->entry, &rq->list);
+		fastlock_release(&rq->lock);
+	} while (1);
+}
+
+int ofi_mr_monitor_get_event(struct ofi_mr_monitor *mr_monitor,
+			     struct ofi_mr_monitor_rq *rq,
+			     struct ofi_mr_monitor_context **context)
+{
+	int ret;
+
+	assert(mr_monitor && context);
+
+	util_mr_monitor_read_events(mr_monitor);
+
+	fastlock_acquire(&rq->lock);
+	if (!dlist_empty(&rq->list)) {
+		dlist_pop_front(&rq->list,
+				struct ofi_mr_monitor_context,
+				*context, entry);
```

Comment:
and here call
```
dlist_init(&subscript->entry);
```
when removing the subscription from the list to avoid double insertions.

Reply:
Done